File: src/utils/encryption.js
import crypto from 'crypto';

const ENCRYPTION_ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;

const getEncryptionKey = () => {
  const secret =
    process.env.CUSTOM_API_ENCRYPTION_KEY ||
    process.env.ENCRYPTION_KEY ||
    process.env.JWT_SECRET;

  if (!secret) {
    throw new Error('Encryption key is not configured');
  }

  return crypto.createHash('sha256').update(secret).digest();
};

export const encrypt = (plainText) => {
  if (!plainText) {
    return null;
  }

  const key = getEncryptionKey();
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ENCRYPTION_ALGORITHM, key, iv);

  const encrypted = Buffer.concat([cipher.update(plainText, 'utf8'), cipher.final()]);
  const authTag = cipher.getAuthTag();

  return `${iv.toString('base64')}:${authTag.toString('base64')}:${encrypted.toString('base64')}`;
};

export const decrypt = (payload) => {
  if (!payload) {
    return null;
  }

  const [ivB64, authTagB64, encryptedB64] = payload.split(':');
  if (!ivB64 || !authTagB64 || !encryptedB64) {
    throw new Error('Invalid encrypted payload format');
  }

  const key = getEncryptionKey();
  const iv = Buffer.from(ivB64, 'base64');
  const authTag = Buffer.from(authTagB64, 'base64');
  const encrypted = Buffer.from(encryptedB64, 'base64');

  const decipher = crypto.createDecipheriv(ENCRYPTION_ALGORITHM, key, iv);
  decipher.setAuthTag(authTag);

  const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);
  return decrypted.toString('utf8');
};

export default {
  encrypt,
  decrypt
};

File: app/utils/encryption.py
"""Encryption utilities."""
import os
import base64
from cryptography.fernet import Fernet
from app.utils.logger import get_logger

logger = get_logger(__name__)

# Initialize Fernet cipher
def _get_encryption_key():
    """Get or generate encryption key."""
    key_env = os.getenv("ENCRYPTION_KEY", "")
    
    if key_env:
        try:
            # Validate that it's a proper Fernet key (base64 encoded 32 bytes)
            key_bytes = key_env.encode() if isinstance(key_env, str) else key_env
            # Test if it's a valid Fernet key
            Fernet(key_bytes)
            return key_bytes
        except Exception as e:
            logger.warning(f"Invalid ENCRYPTION_KEY format: {str(e)}")
            # Fall through to generate new key
    
    # Generate new key if not provided or invalid
    new_key = Fernet.generate_key()
    logger.warning("ENCRYPTION_KEY not set or invalid. Generated temporary key. Set ENCRYPTION_KEY in .env for persistence.")
    return new_key

try:
    ENCRYPTION_KEY = _get_encryption_key()
    CIPHER = Fernet(ENCRYPTION_KEY)
except Exception as e:
    logger.warning(f"Failed to initialize encryption cipher: {str(e)}. Using dummy cipher.")
    # Fallback for testing
    CIPHER = None


def encrypt(plaintext):
    """Encrypt plaintext string."""
    try:
        if not CIPHER:
            logger.warning("Cipher not initialized, returning plaintext")
            return plaintext
        
        if isinstance(plaintext, str):
            plaintext = plaintext.encode()
        encrypted = CIPHER.encrypt(plaintext)
        return encrypted.decode()
    except Exception as e:
        logger.error(f"Encryption error: {str(e)}")
        raise


def decrypt(ciphertext):
    """Decrypt ciphertext string."""
    try:
        if not CIPHER:
            logger.warning("Cipher not initialized, returning ciphertext")
            return ciphertext
        
        if isinstance(ciphertext, str):
            ciphertext = ciphertext.encode()
        decrypted = CIPHER.decrypt(ciphertext)
        return decrypted.decode()
    except Exception as e:
        logger.error(f"Decryption error: {str(e)}")
        raise
